python/lsst/source/injection/inject_coadd.py

```diff
@@ -82,3 +145,175 @@ def runQuantum(self, butler_quantum_context, input_refs, output_refs):
+            self.config.variance_fit_seed_2, self.config.variance_fit_seed_2 + self.config.n_fits_2
+        ):
+            ransac = RANSACRegressor(
+                loss="squared_error",
+                residual_threshold=self.config.threshold_scale_2 * linear_mse,
+                max_trials=self.config.max_trials_2,
+                random_state=seed,
+            )
+            ransac.fit(flux[outlier_mask_1].reshape(-1, 1), var[outlier_mask_1])
+            # Remember fit results
+            slope = ransac.estimator_.coef_[0]
+            fit_results.append((slope, seed))
+
+        # K-Means cluster the second round of fits,
+        # to find the most stable result
+        kmeans = KMeans(
+            n_clusters=self.config.n_clusters_2,
+            random_state=self.config.kmeans_seed_2,
+            n_init=self.config.kmeans_n_init_2,
+        )
+        kmeans.fit(np.log(np.array([f[0] for f in fit_results if f[0] > 0])).reshape(-1, 1))
+        label_counts = [np.sum(kmeans.labels_ == idx) for idx in range(self.config.n_clusters_2)]
+
+        # Recall one of the stable fits
+        stable_fit_seeds = np.array([f[1] for f in fit_results if f[0] > 0])[
+            kmeans.labels_ == np.argmax(label_counts)
+        ]
+        if len(stable_fit_seeds == 0):
+            # No positive-slope fit found.
+            # Allow the fitted slope to be negative but throw a warning.
+            self.log.warning(
+                "No positive-slope result in the second round of "
+                "RANSAC fits. Proceeding with a negative-slope fit."
+            )
+            stable_fit_seeds = np.array([f[1] for f in fit_results])[
+                kmeans.labels_ == np.argmax(label_counts)
+            ]
+        seed = stable_fit_seeds[0]
+        ransac = RANSACRegressor(
+            loss="squared_error",
+            residual_threshold=self.config.threshold_scale_2 * linear_mse,
+            max_trials=self.config.max_trials_2,
+            random_state=seed,
+        )
+        ransac.fit(flux[outlier_mask_1].reshape(-1, 1), var[outlier_mask_1])
+
+        # Pixels with a "good" variance vs. flux relationship:
+        # Union of the inliers from the first fit
+        # together with the inliers from the second fit.
+        flux_good = np.concatenate(
+            (flux[inlier_mask_1], flux[outlier_mask_1][ransac.inlier_mask_]),
+            axis=None,
+        )
+        var_good = np.concatenate(
+            (var[inlier_mask_1], var[outlier_mask_1][ransac.inlier_mask_]),
+            axis=None,
+        )
+
+        # Fit all the good pixels with a simple least squares regression.
+        linear = LinearRegression()
+        linear.fit(flux_good.reshape(-1, 1), var_good)
+        variance_scale = float(linear.coef_[0])
+
+        if variance_scale < 0:
+            self.log.warning("Slope of final variance vs. flux fit is negative.")
+
+        # Return the slope of the final fit.
+        return variance_scale
```

Comment:
If a negative scale factor is returned, what kind of damage could this do? Would it be wise for us to put a hard limit of 0.0 here, or cause a Raise, to prevent this from happening in the wild?

Reply:
No longer relevant following review.